src/hdmf/query.py

```diff
@@ -160,3 +161,30 @@ def __next__(self):
 
     def next(self):
         return self.dataset.next()
+
+
+class ReferenceResolver(with_metaclass(ABCMeta, object)):
+
+    @abstractmethod
+    def invert(self):
+        """
+        Return an object that defers reference resolution
+        but in the opposite direction.
+        """
+        pass
+
+
+class BuilderResolver(ReferenceResolver):
+    """
+    A reference resolver that resolves references to Builders
+    """
+
+    pass
+
+
+class ContainerResolver(ReferenceResolver):
+    """
+    A reference resolver that resolves references to Containers
+    """
+
+    pass
```

Comment:
On the base resolver classes, would it make sense to have 

```
def get_object(self, inobj):
     raise NotImplementedError
```

to define the functions that the subclasses need to implement?

Reply:
I don’t think so, since there’s no guarantee that a backend supports a 1-1 object mapping.